<a href="https://colab.research.google.com/github/Meng-MiamiOH/MTH231/blob/main/exercise_machine_learning_and_coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise for Machine Learning and Coding Session

For this exercise, you will be using the Breast Cancer Wisconsin (Diagnostic) Database to create classifiers that can help diagnose patients. First, read through the description of the dataset (below).

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

print(cancer.DESCR) # Print the data set description

.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean Radius, f

The object returned by load_breast_cancer() is a scikit-learn Bunch object, which is similar to a dictionary.

In [ ]:
cancer.keys()

### 1. Convert the sklearn.dataset `cancer['data']` to a `(569, 30)` DataFrame `X` and convert the `cancer['target']` to a `(569, 1)` Series `y`.

In [ ]:
X =  pd.DataFrame(data=cancer['data'], columns=cancer['feature_names'])
y = pd.Series(cancer['target'])
print(X)
print(y)

### 2. What is the class distribution? (i.e. how many instances of `malignant` (encoded 0) and how many `benign` (encoded 1)?)


In [ ]:
len(y[y == 0])

In [ ]:
len(y[y == 1])

### 4. Using `train_test_split`, split `X` and `y` into training and test sets `X_train`, `X_test`, `y_train`, and `y_test`.

* `X_train` *has shape* `(426, 30)`
* `X_test` *has shape* `(143, 30)`
* `y_train` *has shape* `(426,)`
* `y_test` *has shape* `(143,)`

In [ ]:
# your code

### 5. Using KNeighborsClassifier, fit a k-nearest neighbors (knn) classifier with `X_train`, `y_train` and using one nearest neighbor (`n_neighbors = 1`).

In [ ]:
# your code

### 6. Using your knn classifier, predict the class label using the mean value for each feature.


In [ ]:
mean = X_train.mean().values.reshape(1, -1)
mean

In [ ]:
# your code

### 7. Use the knn classifer your created for Question 5 to predict the class labels for the test set `X_text`.

*The result should be a numpy array with shape `(143,)` and values either `0.0` or `1.0`.*

In [ ]:
# your code

### 8. Find the score (mean accuracy) of your knn classifier using `X_test` and `y_test`.

*The calculated score should be a float between 0 and 1*

In [ ]:
# your code

### 9. Try to train kNN classifier with different k (1, 3, 5, 10) and compare the performance.

In [ ]:
# your code

### 10. Use logistic regression model to create a classifier for the data.

In [ ]:
# your code

### 11. Evaluate the logistic classifier and compare it with kNN classifier.

In [ ]:
# your code